### ISSUE TITLE
Color Override for Websites

### ISSUE THREAD

**Issue by blueto01h** [minbrowser/min#562]
Hey, it would be nice to see a way to override background color and the likes of the displayed websites like in Firefox. Can't really call the dark theme a dark theme if only the tabs etc are dark and white webpages are still blasting the light in your face... It is the only thing that keeps me from using it and sticking to Firefox, otherwise it is a great project and I love the ideas and style.

**Comment by PalmerAL** [comment]
I agree this would have to be a good feature to have, although I'm not really sure how it would work. If we just change the background color of the page to black, a lot of the text ends up unreadable, so we need some way to change the colors on top of it as well. Inverting all of the colors with ```filter: invert(1)``` solves this problem, but then the photographs on the page get messed up.

I guess we should look into what the existing dark theme extensions do; do you happen to know how that works?

**Comment by blueto01h** [comment]
thank you for the answer! unfortunately i do not know how it works. but it is rather crude in firefox as well, so guess it is just changing the background to black, since there is some unreadable text some few times, but this doesnt bother me, i also of course change all text color to grey or something else and if there is something important made unreadable i can simply adjust the colors as i want. its better than nothing :)

**Comment by blueto01h** [comment]
and btw, sometimes i use chromium with an extension which does invert colors but it keeps dark websites still dark. pictures however are sometimes inverted. so it tries to recognize if there is an image and if the color used is already dark.

**Comment by PalmerAL** [comment]
I looked at a couple of dark theme extensions:

<a href="https://chrome.google.com/webstore/detail/dark-theme-everywhere/nibjplbnmmklkfnkpecgbffkifmdbjed">Dark theme everywhere</a> assigns a specific color to every HTML element. This is good because text is usually readable, but it also means that you lose most of the sites own styles, so every site looks the same.

<a href="https://addons.mozilla.org/en-US/firefox/addon/dark-mode-webextension/?src=search">Dark mode webextension</a> does basically the same thing, but it also has built-in exceptions for a bunch of popular websites.

We could probably do this as well; getting it to work consistently and look good everywhere is going to be really difficult, but we could probably make it good enough for most sites.

**Comment by blueto01h** [comment]
that would be really awesome! 
i dont really care too much about the pages original style anyway, as it is a pretty functional thing anyway. some very few times it means loss of information that could be important, but the advantage it has for people with sensitive eyes is over weighing. :)

**Comment by PalmerAL** [comment]
You can get pretty close to achieving this by creating a userscript:

```
var style = document.createElement("style")

style.textContent = `

html, body {
    background: #111 !important;
    color: #fff !important;
}

body * {
    background: #111 !important;
    color: #fff !important;
}

input, button {
border: 1px white solid !important;
}
`

document.head.appendChild(style);
```

To use this, you would need to use the latest version of Min from the master branch (or wait until the 1.8 release), and create a `userscripts` directory inside your data directory, which is  ~/library/application support/min on Mac, \users\username\appdata\roaming\min on Windows, or /home/username/.config/min on Linux. Inside that, you would create a `global.js` file where you would past the script above.

This would need a lot more work before it's reliable enough to include as a built-in option, but it seems to work fairly well.

**Comment by blueto01h** [comment]
awesome! thank you, you're always so quick to answer. i was afk lately a lot, so i was rather slow.

**Comment by blueto01h** [comment]
awesome! thank you,  youre always so quick to answer, unlike me who was afk a lot latey :)
anyways, ive cloned the repository and followed the instrcuctions but in the readme there is only a way to start the app from the repo with electron, but that way i have no directory in .config and i wouldnt know where to place the global.js. 
where can i put it then? starting min from the git repo with electron is problem for me, ill just add an alias for this

**Comment by PalmerAL** [comment]
The directory should be automatically created the first time you run min, regardless of how you start it. For using electron, you can directly type the path to the electron binary as well.

I'm traveling for the next few weeks, so I probably won't respond to this again for a bit.

**Comment by blueto01h** [comment]
awesome, i checked again and this time i found the data directory, but it is in fact starting with a capital letter, all directories inside that are also beginning with a capital letter. 
so i tried anyway to create a directory called "userscripts" with said "global.js" inside and that worked very well, after enabling the userscripts option inside the browsers preferences. 
however everytime a tab is loaded i still get a white flash for a bit, is there a way to fix that in the userscript?

**Comment by PalmerAL** [comment]
Yeah, that is kind of annoying. In order to fix that, we'd have to find a way to inject the script earlier than the `dom-ready` event, I'm not really sure how to do that.

**Comment by IceHacks** [comment]
The developer would have to program that in.

**Comment by PalmerAL** [comment]
@513ry We already support injecting custom scripts starting in 1.8 (see [this comment](https://github.com/minbrowser/min/issues/562#issuecomment-394114985) for instructions). The issue is that the script loads a bit later than we want it to; we need to find a webview event in Electron that allows us to do that.

**Comment by 513ry** [comment]
Ow, sorry, sure you do. I see you are using `dom-ready` to bind events in `js/webview/userscripts.js`. They may be also some event binding of some user scripts with event like `window-onload` for "preload" scripts. But you are probably right, there most be some webview event doing this as well.
Btw. I like your code, it's seems to aim really KISS :).

**Comment by blueto01h** [comment]
OK, thank you anyways, if theres a a way to inject the custom script ealier ill be glad to hear about it. i can also check the code myself and meddle with electron a bit, but in about a month when my exams are finished. :)

**Comment by PalmerAL** [comment]
I've improved the userscript above a bit, and created a repository for it: https://github.com/PalmerAL/min-dark-mode-userscript

For this to become a built-in feature, there are few issues we would need to solve:
* We need to inject the script earlier to avoid the white flash.
* The script needs to handle elements with a CSS background-image.
* Using `filter: invert()` on the entire page seems to make rendering slower and increase power usage by a fair amount, although I'm not sure if this is actually fixable.

**Comment by sturlabragason** [comment]
Here's what I feel does the best job when it comes to achieving what's being requested above: [https://github.com/darkreader/darkreader](https://github.com/darkreader/darkreader).

Maybe a fork could be made that supports min.

**Comment by juanvalag2019** [comment]
Hi,
You can use this user script, it includes what @sturlabragason suggests:
[darkreader.txt](https://github.com/minbrowser/min/files/6844737/darkreader.txt)
